hrs-portlets-ps-impl/src/main/java/edu/wisc/hrs/dao/absbal/DuplicateZeroVacationCarryoverBalanceFilter.java

```diff
@@ -0,0 +1,78 @@
+package edu.wisc.hrs.dao.absbal;
+
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import edu.wisc.hr.dm.absbal.AbsenceBalance;
+import edu.wisc.hr.dm.person.Job;
+import org.apache.commons.lang.Validate;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Filter that removes extra zero-balance vacation carryover balance entitlements,
+ * retaining just one zero-balance vacation carryover balance entitlement per job.
+ */
+public final class DuplicateZeroVacationCarryoverBalanceFilter {
+
+    /**
+     * Predicate that's true for zero-balance "Vacation Carryover Balance" absence balances.
+     */
+    private final Predicate<AbsenceBalance> zeroVacationCarryoverPredicate =
+            Predicates.and(
+                    new VacationCarryoverBalancePredicate(),
+                    new ZeroBalancePredicate()
+            );
+
+    /**
+     * Given a List of AbsenceBalances, return a List of AbsenceBalances with the same content as the input list except
+     * containing no zero-balance "Vacation Carryover Balance" entitlements other than first such zero-balance
+     * "Vacation Carryover Balance" per job. That is, for any given job, an employee should have at most one
+     * zero-balance "Vacation Carryover Balance", and that one should be the first such balance from the input list.
+     * @param input
+     * @return a filtered List<AbsenceBalance>
+     * @throws IllegalArgumentException if input is null
+     */
+    public List<AbsenceBalance> filter(final List<AbsenceBalance> input) {
```

Comment:
CodeClimate is complaining about the "cognitive complexity" of `filter()` (scores 6, worse than the allowable 5).

I don't see how to make it structurally less complex. It's doing what it essentially needs to be doing: keeping around just enough state about what it has seen and deciding to include, exclude, and update that state as it goes.